srv/src/main/java/my/bookshop/handlers/AdminServiceAddressHandler.java

```diff
@@ -0,0 +1,189 @@
+import com.sap.cloud.sdk.cloudplatform.resilience.ResilienceConfiguration.TimeLimiterConfiguration;
+import com.sap.cloud.sdk.cloudplatform.resilience.ResilienceDecorator;
+
+import cds.gen.adminservice.Addresses;
+import cds.gen.adminservice.AdminService_;
+import cds.gen.adminservice.Orders;
+import cds.gen.api_business_partner.ABusinessPartnerAddress;
+import cds.gen.api_business_partner.ABusinessPartnerAddress_;
+import cds.gen.api_business_partner.ApiBusinessPartner_;
+import cds.gen.api_business_partner.BOBusinessPartnerChanged;
+import cds.gen.my.bookshop.Addresses_;
+import my.bookshop.MessageKeys;
+
+/**
+ * Custom handler for the Admin Service Addresses, which come from a remote S/4 System
+ */
+@Component
+@ServiceName(AdminService_.CDS_NAME)
+public class AdminServiceAddressHandler implements EventHandler {
+
+	private final static Logger logger = LoggerFactory.getLogger(AdminServiceAddressHandler.class);
+
+	// We are mashing up the AdminService with two other services...
+	@Autowired
+	private PersistenceService db;
+
+	@Resource(name = ApiBusinessPartner_.CDS_NAME)
+	private CdsService bupa;
+
+	// Using reflected definitions from the model (request-dependant, and therefore tenant-dependant)
+	private Class<Addresses_> addresses = Addresses_.class;
+	private Class<ABusinessPartnerAddress_> externalAddresses = ABusinessPartnerAddress_.class;
+
+	@Autowired
+	private CdsModel model;
+
+	private String[] getRelevantColumns() {
+		// determine columns from simplified Address entity definition
+		return model.getEntity(cds.gen.api_business_partner.Addresses_.CDS_NAME).elements().map(e -> e.getName()).collect(Collectors.toList()).toArray(new String[0]);
+	}
+
+	// Delegate ValueHelp requests to S/4 backend, fetching current user's addresses from there
+	@On(entity = cds.gen.adminservice.Addresses_.CDS_NAME)
+	public Result readAddresses(CdsReadEventContext context) {
+		String businessPartner = context.getUserInfo().getAttributeValues("businessPartner").stream().findFirst()
+			.orElseThrow(() -> new ServiceException(ErrorStatuses.FORBIDDEN, MessageKeys.BUPA_MISSING));
+
+		// forward the existing where condition
+		Select<?> select = Select.from(externalAddresses).where(a -> a.BusinessPartner().eq(businessPartner));
+		Optional<CqnPredicate> where = context.getCqn().where();
+		if(where.isPresent()) {
+			select.where(CQL.and(select.where().get(), where.get()));
+		}
+
+		// forward the selected columns (in Java, we cannot use our simplified projection Addresses yet)
+		List<CqnSelectListItem> columns = context.getCqn().items();
+		boolean hasStar = columns.stream().anyMatch(CqnSelectListItem::isStar);
+		if (hasStar || columns.isEmpty()) {
+			select.columns(getRelevantColumns());
+		} else {
+			List<String> relevantColumnsList = Arrays.asList(getRelevantColumns());
+			select.columns(columns.stream().filter(i -> relevantColumnsList.contains(i.asValue().value().asRef().firstSegment())).collect(Collectors.toList()));
```

Comment:
I'm not sure if we can be sure that `i` is a reference here, but I guess if don't assume this, it would make this logic even more complex..

Reply:
Yes I think this is something we can ignore for now. In the future this code will be completely removed, as soon as we support projections on remote entities. This here is just a poor mans solution for implementing this feature.